node/src/components/block_executor.rs

```diff
@@ -278,9 +278,20 @@ impl BlockExecutor {
             None => {
                 let height = finalized_block.height();
                 debug!("no pre-state hash for height {}", height);
-                // The parent block has not been executed yet; delay handling.
-                self.exec_queue.insert(height, (finalized_block, deploys));
-                Effects::new()
+                // re-check the parent map - the parent might have been executed in the meantime!
+                if let Some(state_root_hash) = self.pre_state_hash(&finalized_block) {
+                    let state = Box::new(State {
+                        finalized_block,
+                        remaining_deploys: deploys,
+                        execution_results: HashMap::new(),
+                        state_root_hash,
+                    });
+                    self.execute_next_deploy_or_create_block(effect_builder, state)
+                } else {
+                    // The parent block has not been executed yet; delay handling.
+                    self.exec_queue.insert(height, (finalized_block, deploys));
+                    Effects::new()
+                }
```

Comment:
I wonder whether it would be cleaner (and less code) to just put it in the queue in `handle_get_deploys_result`, as soon as we _request_ the parent? Then we'd never be in this situation and miss the point where the parent becomes available.

Reply:
But then we risk executing the block twice, don't we? Once on reading the entry from the queue, and once when we get the result from the linear chain (which can return the block).

Reply:
Yes, we'd have to take it out of the queue when we execute it. Anyway, just a thought. It's fine as it is, too.